### PR TITLE
Fix unhealthy nodes processing when wait_primary.

### PR DESCRIPTION
**Pull request by DimCitus**
The FSM state wait_primary embeds the information that there is no solution
to failover, and as such there is no point in changing the state of the node
to draining or otherwise, as no failover process is going to kick-in.

### CODE REVIEW

File: tests/test_multi_async.py [hunk 1]
```diff
@@ -178,3 +178,45 @@ def test_012_ifup_node4():
...
+
+    ssn = ""
+    eq_(node2.get_synchronous_standby_names(), ssn)
+    eq_(node2.get_synchronous_standby_names_local(), ssn)
+
+def test_015_stop_primary():
+    node2.fail()
+
+    print()
+    assert node3.wait_until_state(target_state="report_lsn")
```
Comment (JelteF) [inline tests/test_multi_async.py:204]: I don't think this should happen. It should stay a `secondary` (or should become `catchingup` maybe).
This causes some weird behavior as a side effect. When node2 comes up now, without node1 coming up as well, suddenly node2 becomes `primary` instead of staying `wait_primary`.

Comment (JelteF) [reply tests/test_multi_async.py:204]: This also seems to mess up some other things. When node1 comes up after what I described above, then it will stay in demoted forever.

Comment (DimCitus) [reply tests/test_multi_async.py:204]: Okay I'm going to see about adding another series of tests with node1 restarted first then. I still thing REPORT_LSN is a good assignment for the non-candidate standby when there's no other node available, because that's the state that we have for a standby without a any upstream node at the moment.

File: src/monitor/group_state_machine.c [hunk 2]
```diff
@@ -1028,13 +1082,55 @@ ProceedGroupStateForMSFailover(AutoFailoverNode *activeNode,
...
+	 * The currently selected node might not be marked healthy at this time
+	 * because in REPORT_LSN we shut Postgres down. We still should proceed
+	 * with the previously selected node in that case.
+	 *
+	 * Also pay attention that Postgres might be running again, as reported by
+	 * the pg_autoctl node-active protocol, while our health checks have not
+	 * picked-up on that fact yet. The IsHealthy() check requires both checks
+	 * to be all-green, finding the candidate is okay with only one of those
+	 * being good: we really need to avoid having two candidates at the same
+	 * time.
```
Comment (JelteF) [inline src/monitor/group_state_machine.c:1095]: I believe the changes that this comment addresses have been (re)moved again. I don't see anything related to `isRunning` and the `REPORT_LSN` stuff is addressed below in a second comment already.

File: tests/test_basic_operation.py [hunk 3]
```diff
@@ -75,31 +75,48 @@ def test_005_read_from_secondary():
...
     print("Disabling maintenance on node1")
     node1.disable_maintenance()
     assert node1.wait_until_pg_is_running()
     assert node1.wait_until_state(target_state="secondary")
     assert node2.wait_until_state(target_state="primary")
 
-    eq_(node2.get_synchronous_standby_names_local(), '*')
+    ssn = '*'
+    eq_(node2.get_synchronous_standby_names_local(), ssn)
+    eq_(node2.get_synchronous_standby_names(), ssn)
```
Comment (JelteF) [inline tests/test_basic_operation.py:117]: It seems we're doing this type of check a lot. Maybe add it as a helper method so we can do:
```
node2.check_synchronous_standby_names("*")
```

Can also be done in a separate PR.

File: tests/test_multi_async.py [hunk 4]
```diff
@@ -178,3 +178,112 @@ def test_012_ifup_node4():
...
+
+def test_014_003_restart_node1():
+    node1.run()
+
+    # node1 used to be primary, now demoted, and meanwhile node2 was primary
+    # so we can't decide to promote node1 now, it's stuck demoted
+    assert node1.wait_until_state(target_state="demoted")
+    assert node3.wait_until_state(target_state="report_lsn")
```
Comment (JelteF) [inline tests/test_multi_async.py:230]: I think this needs a bit of a `sleep` and check state again to confirm that `node1` doesn't move away from demoted.

File: tests/test_multi_ifdown.py [hunk 5]
```diff
@@ -240,7 +243,7 @@ def test_015_finalize_failover_after_most_advanced_secondary_gets_back():
 
     # and, node2 should finally become the primary without losing any data
     print()
-    assert node2.wait_until_state(target_state="primary")
```
Comment (JelteF) [inline tests/test_multi_ifdown.py:243]: Why is this change expected?

File: tests/test_multi_async.py [hunk 6]
```diff
@@ -178,3 +178,112 @@ def test_012_ifup_node4():
...
+    node1.fail()
+
+    print()
+    assert node3.wait_until_state(target_state="report_lsn")
+
+def test_015_003_restart_node1():
+    node1.run()
+
+    # node1 used to be primary, now demoted, and meanwhile node2 was primary
+    # so we can't decide to promote node1 now, it's stuck demoted
```
Comment (JelteF) [inline tests/test_multi_async.py:277]: This comment is copy pasted from test 014 and is incorrect.

**Review by JelteF — APPROVED** [review]
Looks good overal, left some small comments.

File: tests/test_basic_operation.py (hunk 3 above)
Comment (DimCitus) [reply tests/test_basic_operation.py:117]: Yeah separate PR would be good. And in the 1.4.2 milestone too. See https://github.com/citusdata/pg_auto_failover/issues/527

File: tests/test_multi_ifdown.py (hunk 5 above)
Comment (DimCitus) [reply tests/test_multi_ifdown.py:243]: We just do the same test later, waiting for the primary node first have an higher chance of reaching a timeout compared to waiting for the secondary nodes first.

File: src/monitor/group_state_machine.c (hunk 2 above)
Comment (DimCitus) [reply src/monitor/group_state_machine.c]: Good point, yeah, this part of the code received several changes following some QA sessions and testing. Fixed now.

File: tests/test_multi_async.py (hunk 4 above)
Comment (DimCitus) [reply tests/test_multi_async.py:233]: Added now, thanks.

File: tests/test_multi_async.py (hunk 6 above)
Comment (DimCitus) [reply tests/test_multi_async.py]: Removed, thanks.

**Review by JelteF — APPROVED** [review]

File: src/monitor/group_state_machine.c [hunk 7]
```diff
@@ -1031,10 +1085,45 @@ ProceedGroupStateForMSFailover(AutoFailoverNode *activeNode,
...
+			message, BUFSIZE,
+			"Active " NODE_FORMAT
+			" found failover candidate " NODE_FORMAT
+			" being promoted, it reported state \"%s\"",
+			NODE_FORMAT_ARGS(activeNode),
+			NODE_FORMAT_ARGS(nodeBeingPromoted),
+			ReplicationStateGetName(nodeBeingPromoted->reportedState));
+
+		/*
+		 *
```
Comment (JelteF) [inline src/monitor/group_state_machine.c:1110]: There was an extra newline in this comment.

```suggestion
```

Comment (DimCitus) [reply src/monitor/group_state_machine.c]: Ooops, sorry about that. I guess that's one way to check the stability of the test suite.

**Review by JelteF — APPROVED** [review]
